Add horizontal resizability to autocomplete popup

sakurai-youhei opened the pull request:
## Summary

This PR adds horizontal resizability to the autocomplete popup. The resized width does not persist, so it returns to the default on every load.

![resizer](https://github.com/elastic/kibana/assets/721858/cd0b686d-685b-47cd-80d8-12d2300d7740)

_^^^ Demonstrated on Chrome Version 124.0.6367.91 (Official Build) (64-bit)._

<details>
<summary>Appearance on Edge Version Version 124.0.2478.51 (Official build) (64-bit)</summary>

![image](https://github.com/elastic/kibana/assets/721858/c2d8ae07-321b-4fa9-965b-2c7a2ee3e89f)
![image](https://github.com/elastic/kibana/assets/721858/ab93af77-7311-4e4b-8300-785023830542)

</details>

<details>
<summary>Appearance on Firefox 125.0.2 (64-bit)</summary>

![image](https://github.com/elastic/kibana/assets/721858/630cec2c-45ee-4894-8bdf-f92cb5dfcbc3)
![image](https://github.com/elastic/kibana/assets/721858/56c62976-abc8-4803-a969-75ff0f8049ae)

</details>

<details>
<summary>Appearance on Safari 17.2.1 (19617.1.17.11.12)</summary>

Show scroll bars=Automatically
<img width="324" alt="2024-04-26 19 10 24" src="https://github.com/elastic/kibana/assets/721858/2a35348a-8637-44ae-9ba0-73e9633f4f26">

Show scroll bars=Always
<img width="324" alt="2024-04-26 19 11 02" src="https://github.com/elastic/kibana/assets/721858/09533ff4-8759-44b2-9395-79cbfb8da392">

</details>

<details>
<summary>Appearance on iOS Safari 17.4.1 - popup does not looks good same as before</summary>

vvvThis branch:
![image](https://github.com/elastic/kibana/assets/721858/eeebc3ac-3112-4a96-9e4f-4ffff7462c3a)

vvv8.13.0:
![image](https://github.com/elastic/kibana/assets/721858/8fa052c2-5277-45c5-8c8b-1a2cc5f5335b)

</details>

<details>
<summary>Appearances at 232897c</summary>

![autocomplete-resizability-chrome-3](https://github.com/elastic/kibana/assets/721858/31772c4b-949c-4cea-b2db-2d19b8b3dd72)
_^^^ Demonstrated on Chrome Version 123.0.6312.123 (Official Build) (64-bit)._

<details>
<summary>Appearance on Edge Version 123.0.2420.81 (Official build) (64-bit)</summary>

![image](https://github.com/elastic/kibana/assets/721858/06f18fdd-9ed8-4b59-a246-c3d50a604dc5)
![image](https://github.com/elastic/kibana/assets/721858/6b11e10b-9e77-49b5-b9a1-13e2840121fb)

</details>

<details>
<summary>Appearance on Firefox 124.0.2 (64-bit)</summary>

![image](https://github.com/elastic/kibana/assets/721858/0512b72e-deb4-4f8e-840f-27005a4f5412)
![image](https://github.com/elastic/kibana/assets/721858/6e44d9b3-f9a6-4933-8fcb-cd2a51e5521b)

</details>

<details>
<summary>Appearance on Safari 17.2.1 (19617.1.17.11.12)</summary>

![image](https://github.com/elastic/kibana/assets/721858/4e228c21-6b45-4544-9bfc-aa8aeb8a6d25)
![image](https://github.com/elastic/kibana/assets/721858/eeea6de0-684a-43a8-ba7b-75d7a289633f)

</details>
</details>

<details>
<summary>Appearances at 98864be</summary>

![autocomplete-resizability-2](https://github.com/elastic/kibana/assets/721858/547b12c8-87d6-43ea-8eb4-e651a9e59b11)
_^^^ Demonstrated on Chrome Version 123.0.6312.106 (Official Build) (64-bit)._

<details>
<summary>Appearance on Edge Version 123.0.2420.81 (Official build) (64-bit)</summary>

![autocomplete-resizability-edge-2](https://github.com/elastic/kibana/assets/721858/d1a1b584-0fed-4688-abaa-8c1bdfb0cee1)
</details>

<details>
<summary>Appearance on Firefox 124.0.2 (64-bit)</summary>

![autocomplete-resizability-firefox-2](https://github.com/elastic/kibana/assets/721858/7c3ac83b-7f2b-4370-a60b-fdcfea7dd40a)
</details>

<details>
<summary>Appearance on Safari 17.2.1 (19617.1.17.11.12)</summary>

![image](https://github.com/elastic/kibana/assets/721858/dd1bcd0d-b02a-4a0a-9b1a-02feec6b68a6)
</details>
</details>

Closes #171268
Closes #125186

### Justification

The autocomplete popup width has been fixed to `280px` [HERE](https://github.com/thlorenz/brace/blob/v0.11.1/ext/language_tools.js#L1209) for a long time. Extending it would also be appreciated as [ace 1.32.9](https://github.com/ajaxorg/ace/blob/v1.32.9/src/autocomplete/popup.js#L449) defaults it to `300px`, and other projects using the ace make it even wider. However, this PR doesn't scope the extension because it's not easy to determine the best width. Furthermore, automatic fitting to suggestions' width is also excluded from the scope to avoid generating new corner cases.

A horizontal scrollbar is an alternative solution, but it's less cost-effective because it requires writing code to sync two separate DOM elements,  `.ace_scrollbar.ace_scrollbar-h` and `.ace_scroller`.

For keyboard accessibility, it's definitely best to add two shortcuts, making the autocomplete popup wider and narrower. However, the questions are what key combinations are natural for them and what delta is appropriate per expansion/shrink. This could be addressed as another enhancement **if key combinations were suggested** at least.

### CSS resizer on Safari

The CSS resizer behaves uniquely on Safari. You can't resize an element if another element overlaps the resizer like this.

```html
data:text/html,<div%20style="position:fixed;width:280px;height:128px;resize:both;border:solid%201px;overflow:hidden;"><div%20style="position:absolute;right:0px;bottom:0px;">FOO<div></div>
```

https://github.com/elastic/kibana/assets/721858/3101b8c2-78b6-4587-8ff1-ed4d6035a3be

The workaround is to set a negative z-index to the overlapping element like this.

```html
data:text/html,<div%20style="position:fixed;width:280px;height:128px;resize:both;border:solid%201px;overflow:hidden;"><div%20style="position:absolute;right:0px;bottom:0px;z-index:-1;">FOO<div></div>
```

### Checklist

- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

## Release note

Adds horizontal resizability to the autocomplete popup on the Dev Tools Console

ElenaStoeva reviewed (commented):
Thanks for working on this @sakurai-youhei. I tested locally and the resizability works well overall. However, I noticed a small visual regression - now the side bar covers the last letter of each word specifying the type of the suggestion:
<img width="389" alt="Screenshot 2024-04-12 at 18 51 00" src="https://github.com/elastic/kibana/assets/59341489/deeb34fb-5fc5-477d-821d-f9940d32acfc">
You can uncover them by extending the autocomplete popup to the right, but they are covered in the initial state when the popup appears and I don't think this looks good visually. Would there be any fix for this?

Also, please keep in mind that we are in the process of migrating Console to Monaco editor (see https://github.com/elastic/kibana/issues/180207) so we might need to implement this there as well once we complete the migration.

sakurai-youhei commented:
@ElenaStoeva Thank you for finding the regression. I didn't notice it because it didn't impact Chrome on my laptop. I've fixed it in the latest commit.

Regarding the new Console, noted and will try to implement it somehow through another PR.

ElenaStoeva reviewed (approved):
Thanks for addressing my feedback @sakurai-youhei! Tested again and now it looks good to me.

> Regarding the new Console, Monaco Editor has built-in horizontal and vertical resizers that will be activated by one or more suggestions even in the new Console. The rest is definitely off-topic, but the following is a quick implementation of the suggestionProvider and how it looks.

Noted, thanks! We'll keep this in mind when we implement the autocomplete functionality in the Console Monaco editor. cc: @yuliacech

MichaelMarcialis commented on src/plugins/console/public/styles/_index.scss:
```diff
@@ -7,3 +7,19 @@
+  .ace_scroller {
+    right: 6px !important;
+
+    .ace_rightAlignedText {
+      right: 10px;
+    }
```
Am I correct in assuming this style was added to prevent the potential for part of the right-aligned text overflowing out of the container (as reported by @ElenaStoeva)? If so, I'm not sure this fully addresses the issue. There is still an overflow issue that still presents itself when shrinking the autocomplete element horizontally. 

![CleanShot 2024-04-15 at 14 57 12](https://github.com/elastic/kibana/assets/3884767/ac5aa53f-f918-4d7b-840b-e56ded75df96)

I imagine this can be addressed more holistically by either placing a minimum width on the element or styling it so that the contexts within can shrink and truncate at smaller sizes.

MichaelMarcialis replied:
In my testing across the latest versions of Chrome, Safari, Opera, Firefox, and Edge, removing this style does not cause the cutting off of the right-aligned "endpoint" text that you described. My guess is that issue was observed because the panel was resized to be slightly narrower than its default width, which then causes content to be cut-off and overflow out of the container (as the GIF above demonstrates). As this style doesn't address the issue, I would suggest removing it altogether.

As for addressing the issue of the panel contents not adapting to the width when the panel is reduced in width, if it's too difficult to correct, I suppose it would be OK to postpone the issue for now, as it sounds like we're switching from Ace to Monaco soon. We can revisit enhancements at that point.

sakurai-youhei replied:
Thank you very much for checking this in detail. This PR no longer contains the `right` adjustments that change visual alignments because I've found an alternative to expose the resizer on Safari. So, I mark this as resolved.

MichaelMarcialis reviewed (commented):
Thanks @sakurai-youhei! I've added some comments below for my initial review.

sakurai-youhei commented on src/plugins/console/public/styles/_index.scss:
```diff
@@ -7,3 +7,19 @@
+  }
+
+  .ace_scrollbar.ace_scrollbar-v {
+    bottom: 6px !important;
```
@MichaelMarcialis Thank you for your review. Indeed, 6px may be too small. I follow the idea.

sakurai-youhei commented on src/plugins/console/public/styles/_index.scss:
```diff
@@ -7,3 +7,19 @@
+  resize: horizontal;
+
+  .ace_scroller {
+    right: 6px !important;
```
This seems necessary to me. Without this, `.ace_scroller` element covers and hides the right edge of `.ace_editor.ace_autocomplete` element on Safari and Firefox. Firefox seems to allow users to click a resizer behind elements, but Safari does not. So, it seems at the moment that the default styling makes the resizer clickable on Chrome, Edge, and Firefox but not Safari.

MichaelMarcialis replied:
In my testing in the latest versions of Chrome, Firefox, Safari, Opera, and Edge, removing this style doesn't appear to cause any issues (in fact, it looks better as it removes the aforementioned text collisions). I'm still able to resize the panel in Safari. From what I can see, it's the styles on lines 22–25 below that ensure the panel is resizable in Safari. Assuming I'm not missing something, my suggestion would be to remove this.

sakurai-youhei replied:
@MichaelMarcialis Thank you very much for the clarification. That sounds like my Safari 17.2.1 behaves differently from yours (17.4.x?). Would you mind if I ask you to check the following reproducer?

```html
data:text/html,<div%20style="position:fixed;width:280px;height:128px;resize:both;border:solid%201px;overflow:hidden;"><div%20style="position:absolute;right:0px;bottom:0px;">FOO<div></div>
```

I can resize the parent div element on my Chrome, Firefox, and Edge but can't on my Safari 17.2.1 due to the overlapping child div element like this.

https://github.com/elastic/kibana/assets/721858/3101b8c2-78b6-4587-8ff1-ed4d6035a3be

Because of this problem, I had to add this style. Without this style, the same problem prevents the resizer from being active on Safari 17.2.1 like this.

https://github.com/elastic/kibana/assets/721858/72147604-28d6-45fa-a57c-a8a05427501f

MichaelMarcialis replied:
@sakurai-youhei: Yes, I'm on Safari 17.4.1. Despite that, my Safari functions exactly the same as yours, and the inability to grab the resize handle can be reproduced using the example code you provided above. However, I do not see the same when testing locally after commenting out lines 14–20 in `_index.scss`. Instead, I see the following:

![CleanShot 2024-04-22 at 10 08 51](https://github.com/elastic/kibana/assets/3884767/ff07c11f-ce7a-47a0-a3b5-39ca91224e0d)

![CleanShot 2024-04-22 at 10 10 34](https://github.com/elastic/kibana/assets/3884767/b05f0845-1424-4080-a5de-9eaa303b0172)

The reason it still works is because there are default inline styles applied to `.ace_scroller`, as seen in the inspector below. When they aren't being overridden by us, everything works fine in Safari. I suggest we just roll with those defaults.

![CleanShot 2024-04-22 at 10 10 34](https://github.com/elastic/kibana/assets/3884767/728b0e6c-0d68-47f5-a027-80925053bdd0)

sakurai-youhei replied:
@MichaelMarcialis Thank you for the detailed information.

Thanks to that, I find that the [Show scroll bars](https://support.apple.com/guide/mac-help/change-appearance-settings-mchlp1225/mac) setting makes a difference [HERE](https://github.com/thlorenz/brace/blob/v0.11.1/index.js#L1084-L1119). The `Automatically` leads 0 while the `Always` leads 15, and the calculated scroll bar width is propagated to the `right` inline style on `.ace_scroller`. What a beautiful trap! :)

So now I can start thinking of the rest. Once I get some idea, I will prepare and push a commit dealing with this point.

sakurai-youhei replied:
Finally, I have found this style is unnecessary thanks to you. I appreciate your time on this very much!

MichaelMarcialis reviewed (approved):
Thanks for the change and comments @sakurai-youhei. I've responded to your comments below. Assuming you're able to address those responses, I'll go ahead and approve this now to unblock you for feature freeze.

sakurai-youhei commented:
@MichaelMarcialis I changed how to expose the resizer on Safari and updated the PR description accordingly.

I believe the negative z-index doesn't create a new incompatibility because https://github.com/elastic/kibana/pull/151386 has already introduced it. But I know it doesn't mean zero concerns around this negative z-index trick. Although I've gone through browsers again, I appreciate your review & checks again. If you find this way is not preferable, please let me know, and I will try to find another.

MichaelMarcialis reviewed (approved):
LGTM. Thanks!
